backend/src/util/githubUtil.ts

```diff
@@ -0,0 +1,310 @@
+  createdAt: number;
+  content: string;
+};
+
+type ReviewedPR = {
+  url: string;
+  createdBy: string;
+  comments: Comment[];
+};
+
+type OpenedPR = {
+  url: string;
+  createdBy: string;
+  createdAt: number;
+  diffSize: number;
+};
+
+// useful to see reason why if not valid
+type ValidationResult = {
+  status: 'valid' | 'invalid' | 'pending';
+  reason?: string;
+};
+
+/** Parses GitHub PR `url` for information necessary to make API calls.
+ *  Raises an error if URL is malformed. */
+const parseGithubUrl = (url: string): PullRequest => {
+  // of the form: https://github.com/cornell-dti/idol/pull/266
+  const pattern = /.*github.com\/([_a-zA-Z0-9-]+)\/([_a-zA-Z0-9-]+)\/pull\/([0-9]+)/;
+
+  const match = url.match(pattern);
+  if (match == null) {
+    throw new Error(`Malformed URL ${url}.`); // more helpful error message?
+  }
+
+  return { owner: match[1], repo: match[2], pull_number: parseInt(match[3], 10) };
+};
+
+/** Returns GitHub username from profile `url`.
+ *  Raises an error if URL is malformed. */
+const parseGithubUsername = (url: string): string => {
+  // of the form: https://github.com/JacksonStaniec
+  const pattern = /.*github.com\/([_a-zA-Z0-9-]+).*/;
+
+  const match = url.match(pattern);
+  if (match == null) {
+    throw new Error(`Malformed URL ${url}.`);
+  }
+
+  return match[1];
+};
+
+/** Returns tuple of earliest valid time, deadline, and Idol member github username.
+ *  Raises an error if the Idol member from `submission` does not have a github username. */
+const parsePortfolioSubmission = (portfolio: DevPortfolio, submission: DevPortfolioSubmission) => {
+  const start = portfolio.earliestValidDate;
+  const end = portfolio.deadline;
+  const usernameUrl = submission.member.github;
+
+  // check github user
+  if (!usernameUrl) {
+    const name = `${submission.member.firstName} ${submission.member.lastName}`;
+    const netid = `${submission.member.netid}`;
+
+    throw new Error(`Idol member ${name} (${netid}) does not have a github username.`);
+  }
+
+  const username = parseGithubUsername(usernameUrl);
+
+  return { start, end, username };
+};
+
+/** Retrieves review comments made on `pull`. */
+const getReviewComments = async (pull: PullRequest): Promise<Comment[]> => {
+  const octokit = new Octokit();
+
+  return Promise.all([
+    octokit.rest.pulls.listReviewComments(pull),
+    octokit.rest.pulls.listReviews(pull)
+  ]).then(([reviewCommentsRes, approvalCommentsRes]) => {
+    // extract information necessary to make a Comment object
+    // works for both review comments and approval comments
+    const toComment = (data): Comment => ({
+      commentUrl: data.html_url,
+      createdBy: data.user.login,
+      createdAt: Date.parse(data.created_at) || Date.parse(data.submitted_at),
+      content: data.body
+    });
+
+    const reviewComments = reviewCommentsRes.data;
+    const approvalComments = approvalCommentsRes.data.filter((comment) => comment.body !== ''); // most are empty
```

Comment:
There is yet another different type of comment that requires a separate API call: the comments attached to PR approvals/rejections. (Henry's "LGTM!" comment on [this PR](https://github.com/cornell-dti/idol/pull/268) for example.)

Review comments are classified as a type of approval/rejection that have a status of "COMMENTED" (as opposed to "APPROVED" or "REJECTED"), but they have a blank body for some reason. So, most of the responses from this second API call are useless, so I filter them first.

Reply:
The goal is to get every comment made by the user under a PR, and consider them all together as a "review." So, stuff like "LGTM" is considered for the review requirement. However, with PRs like [this](https://github.com/cornell-dti/idol/pull/268), the sole "LGTM" comment doesn't satisfy the minimum length requirement, and so this is a "trivial" review. I think it is good to include these approval comments because most times they are just "LGTM" or something short, so they don't contribute much to a review, but it could also possibly be something constructive about future PRs to make, things to improve, etc.